flowpm/raytracing.py

```diff
@@ -0,0 +1,236 @@
+        lens_plane = tf.zeros([batch_size, plane_size, plane_size])
+        
+        # Convert coordinates to angular coords, and then into plane coords
+        xy = (xy / tf.expand_dims(d,-1))/np.pi*180*60/plane_resolution
+        xy = xy + plane_size/2
+        
+        # Selecting only the particles contributing to the lens plane
+        mask = tf.where((d>(nz - (i+1)*w)) & (d <= (nz - i*w)),1.,0.)
+        # And falling inside the plane, NOTE: This is only necessary on CPU, on GPU
+        # cic paint 2d can be made to work with non periodic conditions.
+        mask = mask * tf.where((xy[...,0]>0) & (xy[...,0]<plane_size),1.,0.)
+        mask = mask * tf.where((xy[...,1]>0) & (xy[...,1]<plane_size),1.,0.)
+        # Compute lens planes by projecting particles
+        lens_plane = flowpm.utils.cic_paint_2d(lens_plane, xy + plane_size/2 ,mask)
+        lps.append(lens_plane)
+        lps_a.append(ah)
+        
+        # Here we could trim the state vector for particles originally beyond the current lens plane
+        # This way the simulation becomes smaller as it runs and we save resources
+        state = tf.reshape(state, [3,batch_size, nc[0], nc[1],-1, 3])
+        state = state[:,:,:,:,:(nz - i*w - w // 2),:] # We keep w/2 to be safe, so we allow particle to travel
+                                                     # A max distance of width/2
+        # redefine shape of state
+        nc = state.get_shape()[2:5]
+        state = tf.reshape(state, [3,batch_size,-1,3])
+        # So this seems to work, but we should be a tiny bit careful because we break periodicity in the z
+        # direction at z=0.... probably not a big deal but still gotta check what that does.
+        
+        
+        # Force
+        state = force(state, nc, pm_nc_factor=pm_nc_factor, cosmology=cosmology)
+        f = a1
+
+        # Kick again
+        state = kick(state, p, f, a1, cosmology=cosmology)
+        p = a1
+
+    return state, lps_a, lps
+
+
+def cons(cosmo):
+    """
+    Redshift independent prefactor from Poisson equation
+    
+    Parameters:
+    -----------
+    cosmo: Cosmology
+       Cosmological parameters structure
+
+    Returns
+    -------
+    cons: float
+      prefactor from Poisson equation
+     
+    """ 
+    return 3/2*cosmo['Omega0_m']*(cosmo['H0']/constants.c)**2
+
+def nbar_(nc,Boxsize):
+    """
+    mean 3D particle density
+    
+    Parameters:
+    -----------
+    nc : list
+        Size of the cube, number of cells
+    
+    Boxsize: list
+        Physical size of the cube
+
+    Returns
+    -------
+    nbar: float
+     mean 3D particle density
+     
+    """
+    return np.prod(nc)/np.prod(Boxsize)
+
+def A(nc_xy):
+    """
+    2D mesh area in rad^2 per pixel
+    
+    Parameters:
+    -----------
+    nc_xy : int
+       Number of cell for x and  y 
+        
+    Returns
+    -------
+    A:  float
+     2D mesh area in rad^2 per pixel
+     
+    """
+    return ((5*np.pi/180/nc_xy)**2)
```

Comment:
if I understand correcly, here the 5 is for the 5 deg of the simulated field, we should probably leave that as a parameter of the function